qlight/client_cache.go

```diff
@@ -0,0 +1,78 @@
+package qlight
+
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/rawdb"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/private"
+	"github.com/ethereum/go-ethereum/private/cache"
+	"github.com/ethereum/go-ethereum/private/engine/qlightptm"
+	gocache "github.com/patrickmn/go-cache"
+)
+
+type clientCache struct {
+	cacheWithEmpty    CacheWithEmpty
+	privateBlockCache *gocache.Cache
+	db                ethdb.Database
+}
+
+func NewClientCache(db ethdb.Database) (PrivateClientCache, error) {
+	cachingTXManager, ok := private.P.(*qlightptm.CachingProxyTxManager)
+	if !ok {
+		return nil, fmt.Errorf("unable to initialize cacheWithEmpty")
+	}
+	return NewClientCacheWithEmpty(db, cachingTXManager, gocache.New(cache.DefaultExpiration, cache.CleanupInterval))
+}
+
+func NewClientCacheWithEmpty(db ethdb.Database, cacheWithEmpty CacheWithEmpty, gocache *gocache.Cache) (PrivateClientCache, error) {
+	return &clientCache{
+		cacheWithEmpty:    cacheWithEmpty,
+		privateBlockCache: gocache,
+		db:                db,
+	}, nil
+}
+
+func (c *clientCache) AddPrivateBlock(blockPrivateData BlockPrivateData) error {
+	for _, pvtTx := range blockPrivateData.PrivateTransactions {
+		if err := c.cacheWithEmpty.Cache(pvtTx.ToCachable()); err != nil {
+			return err
+		}
+	}
+	if !common.EmptyHash(blockPrivateData.PrivateStateRoot) {
+		return c.privateBlockCache.Add(blockPrivateData.BlockHash.ToBase64(), blockPrivateData.PrivateStateRoot.ToBase64(), gocache.DefaultExpiration)
+	}
+	return nil
+}
+
+func (c *clientCache) CheckAndAddEmptyEntry(hash common.EncryptedPayloadHash) {
```

Comment:
empty is a field of the entry, set to true
```suggestion
func (c *clientCache) CheckEntryThenSetEmpty(hash common.EncryptedPayloadHash) {
```

Reply:
From the `cache.Set` docs: Add an item to the cache, replacing any existing item. 
I think add is better in this case.